### PR TITLE
Don't try to update counter cache when foreign key value is null.

### LINKED ISSUE #14276: CakePHP 4.0.3 CounterCache on null association
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 4.0.3
* Platform and Target: N/A

### What you did
UsersTable has Behavior CounterCache with ['Tenants' => ['user_count']]. Save a user with a tenant_id = null

### What happened
Exception is thrown `Expression is missing operator (IS, IS NOT) with null value.`

### What you expected to happen
No exception to be thrown.

### PR DESCRIPTION
**Pull request by ADmad**
Closes #14276

### CODE REVIEW

File: src/ORM/Behavior/CounterCacheBehavior.php [hunk 1]
```diff
@@ -219,11 +219,24 @@ protected function _processAssociation(
         array $settings
     ): void {
         $foreignKeys = (array)$assoc->getForeignKey();
-        $primaryKeys = (array)$assoc->getBindingKey();
         $countConditions = $entity->extract($foreignKeys);
+        $allNulls = true;
+        foreach ($countConditions as $field => $value) {
```
Comment (ADmad) [inline src/ORM/Behavior/CounterCacheBehavior.php]: This loop is not pretty but it's to account for multi field foreign key with a mix of null and non null values.

File: tests/TestCase/ORM/Behavior/CounterCacheBehaviorTest.php [hunk 2]
```diff
@@ -175,6 +175,27 @@ public function testAddScope()
...
+            'Users' => [
+                'comment_count',
+            ],
+        ]);
+
+        $entity = new Entity([
+            'title' => 'Orphan comment',
+            'user_id' => null,
+        ]);
+        $this->comment->save($entity);
```
Comment (dereuromark) [inline tests/TestCase/ORM/Behavior/CounterCacheBehaviorTest.php]: Should this be saveOrFail? Or the assert maybe checks if saved.

Comment (ADmad) [reply tests/TestCase/ORM/Behavior/CounterCacheBehaviorTest.php]: Without this fix an exception was thrown so no exception is a test enough.

Comment (dereuromark) [reply tests/TestCase/ORM/Behavior/CounterCacheBehaviorTest.php]: Sure,  but the fake assert could still be a useful one maybe.

Comment (dereuromark) [reply tests/TestCase/ORM/Behavior/CounterCacheBehaviorTest.php]: Imagine an (accidental) regression where no exception is being thrown still, but now also not saved anymore. Thus my concern.

**Comment by chopstik** [comment]
Hi @dereuromark @ADmad had the same issue and for the most part the commit/version change resolved the issue on the CounterCache behaviour. I still see the same Exception (IS/IS NOT NULL) when using a finder method on the count field. See below:
```
$this->addBehavior('CounterCache', [
	'ContractorCompanies' => [
		'threshold_jobs_count' => [
			'finder' => 'ThresholdJobsCount',
		],
	],
	'Instructions'        => ['jobs_count'],
]);
```
```
public function findThresholdJobsCount(Query $query, array $options)
{
	$query->andWhere(['Jobs.assigned' => true]);
	$currentWeek = Time::parse('monday this week');
	$query->andWhere(
		$query->newExpr()->gte('Jobs.created', $currentWeek->format('Y-m-d H:i:s'))
	);

	return $query;
}
```
Can you see a way to adjust [CounterCacheBehaviour.php](https://github.com/cakephp/cakephp/pull/14279/files/4fe9a9610db72aa487e7a2a00afda6c9e2331dbe#diff-0795798ecd11b7b1d2151551def1a027) to take a finder method into account?

**Comment by othercorey** [comment]
@chopstik Please create a new issue that references this one.

**Comment by chopstik** [comment]
Will do.

**Comment by chopstik** [comment]
Opened issue #14362

**Pull request merged**